dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/event/mapper/EnrollmentCriteriaMapper.java

```diff
@@ -0,0 +1,200 @@
+    {
+        ProgramInstanceQueryParams params = new ProgramInstanceQueryParams();
+
+        Set<OrganisationUnit> possibleSearchOrgUnits = new HashSet<>();
+
+        User user = currentUserService.getCurrentUser();
+
+        if ( user != null )
+        {
+            possibleSearchOrgUnits = user.getTeiSearchOrganisationUnitsWithFallback();
+        }
+
+        if ( ou != null )
+        {
+            for ( String orgUnit : ou )
+            {
+                OrganisationUnit organisationUnit = organisationUnitService.getOrganisationUnit( orgUnit );
+
+                if ( organisationUnit == null )
+                {
+                    throw new IllegalQueryException( "Organisation unit does not exist: " + orgUnit );
+                }
+
+                if ( !organisationUnitService.isInUserHierarchy( organisationUnit.getUid(), possibleSearchOrgUnits ) )
+                {
+                    throw new IllegalQueryException( "Organisation unit is not part of the search scope: " + orgUnit );
+                }
+
+                params.getOrganisationUnits().add( organisationUnit );
+            }
+        }
+
+        Program pr = program != null ? programService.getProgram( program ) : null;
+
+        if ( program != null && pr == null )
+        {
+            throw new IllegalQueryException( "Program does not exist: " + program );
+        }
+
+        TrackedEntityType te = trackedEntityType != null
+            ? trackedEntityTypeService.getTrackedEntityType( trackedEntityType )
+            : null;
+
+        if ( trackedEntityType != null && te == null )
+        {
+            throw new IllegalQueryException( "Tracked entity does not exist: " + program );
+        }
+
+        TrackedEntityInstance tei = trackedEntityInstance != null
+            ? trackedEntityInstanceService.getTrackedEntityInstance( trackedEntityInstance )
+            : null;
+
+        if ( trackedEntityInstance != null && tei == null )
+        {
+            throw new IllegalQueryException( "Tracked entity instance does not exist: " + program );
+        }
+
+        params.setProgram( pr );
+        params.setProgramStatus( programStatus );
+        params.setFollowUp( followUp );
+        params.setLastUpdated( lastUpdated );
+        params.setLastUpdatedDuration( lastUpdatedDuration );
+        params.setProgramStartDate( programStartDate );
+        params.setProgramEndDate( programEndDate );
+        params.setTrackedEntityType( te );
+        params.setTrackedEntityInstanceUid(
+            Optional.ofNullable( tei ).map( BaseIdentifiableObject::getUid ).orElse( null ) );
+        params.setOrganisationUnitMode( ouMode );
+        params.setPage( page );
+        params.setPageSize( pageSize );
+        params.setTotalPages( totalPages );
+        params.setSkipPaging( skipPaging );
+        params.setIncludeDeleted( includeDeleted );
+        params.setUser( user );
+
+        return params;
+    }
+
+    @Transactional( readOnly = true )
```

Comment:
I wonder the same, the point is that at the moment "`getFromUrl`" is just a placeholder to convert criteria by invoking the old method. Old method was already annotated with r`eadOnly=true`, so since this is invoked within the same class, it wouldn't be accessed by spring proxy. So to stay 100% compatible with already existing code,` readOnly=true `should be left.
Question is if we need it at all.

Reply:
👍